shared/lib/network-utils.ts

```diff
@@ -0,0 +1,89 @@
+import { type WellknownChain } from 'eth-chainlist';
+import { useSelector } from 'react-redux';
+// eslint-disable-next-line import/no-restricted-paths
+import { useSafeChainsListValidationSelector } from '../../ui/selectors';
+import { DAY } from '../constants/time';
+import fetchWithCache from './fetch-with-cache';
+import { getStorageItem, setStorageItem } from './storage-helpers';
+
+export type { WellknownChain } from 'eth-chainlist';
+const CHAIN_SPEC_URL = 'https://chainid.network/chains.json';
+export const cacheKey = `cachedFetch:${CHAIN_SPEC_URL}`;
+
+/**
+ * Requests for the chainlist need to wait for the pre-seed to finish
+ * before they can be used, otherwise it theoretically can return an empty list,
+ * when we do actually have the data already.
+ */
+let preSeedPromise: Promise<void> | null = null;
+
+/**
+ * Retrieve well-known chains from the cache.
+ */
+async function getWellknownChainsFromCache(): Promise<WellknownChain[] | null> {
+  const cachedResponse = (await getStorageItem(cacheKey)) || {};
+  if (cachedResponse) {
+    return cachedResponse;
+  }
+  return null;
+}
+
+/**
+ * Returns a list of well-known chains either from the cache (which might be
+ * a build-time compiled list of networks), or from an up-to-date remote network
+ * list of well-known chains (if the user's `useSafeChainsListValidation`
+ * security setting allows for it).
+ */
+export async function getWellknownChains(): Promise<WellknownChain[]> {
+  await preSeedPromise;
```

Comment:
Previously, we didn't explicitly wait for the cache seeding code to complete before firing `getWellknownChains`. _In practice_ this is not an issue, since the onboarding screen will always take longer than the `preSeedWellknownChains` function call. This is _technically_ a race condition, and since everything is so nicely centralized in this file now I thought it couldn't hurt to add this in.